Optimize presigning for replay.json

ikreymer opened the pull request:
Fixes #2515.

This PR introduces a significantly optimized logic for presigning URLs for crawls and collections.
- For collections, the files needed from all crawls are looked up, and then the 'presign_urls' table is merged in one pass, resulting in a unified iterator containing files and presign urls for those files.
- For crawls, the presign URLs are also looked up once, and the same iterator is used for a single crawl with passed in list of CrawlFiles
- URLs that are already signed are added to the return list.
- For any remaining URLs to be signed, a bulk presigning function is added, which shares an HTTP connection and signing 8 files in parallels (customizable via helm chart, though may not be needed). This function is used to call the presigning API in parallel.


Testing:
1. Call clear presigned API endpoint to clear presigned.
2. Test by loading crawl or collection with large number of files (eg. Vice crawl).
3. Repeat with this branch and current main
4. Collection and crawls should load an order of magnitude faster with this approach.

ikreymer committed:
optimize generating collection replay.json by performing bulk presigned url lookup

ikreymer committed:
sign in parallel

ikreymer committed:
insert in batches

ikreymer committed:
test out all coll

ikreymer committed:
typo

ikreymer committed:
fix

ikreymer committed:
cleanup

ikreymer committed:
try increasing batch size

ikreymer committed:
fix

ikreymer committed:
presign batch size configurable

ikreymer committed:
refactor: support bulk presign for single crawls as well:
- move bulk presign query to basecrawls
- support bulk presign for crawl as well as collection replay

ikreymer committed:
rename

ikreymer committed:
tweaks

ikreymer committed:
catch and ignore bulkwriteerrors where the err is duplicate (data is added anyway, but exception is emitted)
also catch error when presigned_urls index can't be created due to existing index with different duration,
if so, drop and recreate index

ikreymer commented:
The insert_many() call could sometimes throw a duplicate item exception, even though the bulk write succeeds.
Handled by ignoring the error as suggested in: https://stackoverflow.com/questions/44838280/how-to-ignore-duplicate-key-errors-safely-using-insert-many/44838740#44838740

Also added the same for the other insert_many() in readd pages.

ikreymer commented:
Also index creation can fail if the expire interval changes, just readding index in that case.

ikreymer committed:
drop all indexes to recreate

tw4l commented on backend/btrixcloud/basecrawls.py:
```diff
@@ -464,29 +476,118 @@ async def resolve_signed_urls(
+            names = [file["filename"] for file in sign_files]
+
+            first_file = CrawlFile(**sign_files[0])
+            s3storage = self.storage_ops.get_org_storage_by_ref(org, first_file.storage)
```
If we implement per-file custom storage soon, this assumption won't necessarily hold anymore. Not sure whether to optimize here or add it to the custom storage scope as something to address.

ikreymer replied:
Yes, updated in: https://github.com/webrecorder/browsertrix/commit/ff7cbde507f867a0f63871ad0d5044e00fd39a03 to group by StorageRef, and then presign the group. Should behave same way if a single storage, but also support presigning if multiple different StorageRefs

tw4l commented on backend/btrixcloud/colls.py:
```diff
@@ -554,32 +560,24 @@ async def list_collections(
-        resources = []
-        pages_optimized = True
+        if coll_id:
+            crawl_ids = await self.get_collection_crawl_ids(coll_id)
+            match = {"_id": {"$in": crawl_ids}}
```
```suggestion
            _ = await self.get_collection_raw(coll_id)
            crawl_ids = await self.get_collection_crawl_ids(coll_id)
            match = {"_id": {"$in": crawl_ids}}
```

Might be worth keeping this check in the if coll_id conditional below to avoid the subsequent queries and raise a 404 if the collection doesn't exist

ikreymer replied:
The places that call this already do call `get_collection_raw`, so thought it was redundant..

ikreymer replied:
Will be further addressed by collection org checking as well..

tw4l commented on backend/btrixcloud/colls.py:
```diff
@@ -1009,24 +1007,10 @@ async def list_collection_all(
     @app.get(
         "/orgs/{oid}/collections/$all",
         tags=["collections"],
-        response_model=Dict[str, List[CrawlFileOut]],
```
Should re-add or update response_model

ikreymer replied:
Updated in: https://github.com/webrecorder/browsertrix/pull/2516/commits/b08309eed52804a9877de8e9ed1c4100023fee7b

tw4l commented on backend/btrixcloud/storages.py:
```diff
@@ -521,6 +531,73 @@ async def get_presigned_url(
+                and s3storage.access_endpoint_url != s3storage.endpoint_url
+            ):
+                parts = urlsplit(s3storage.endpoint_url)
+                host_endpoint_url = f"{parts.scheme}://{bucket}.{parts.netloc}/{key}"
```
Might require a change after merging s3v4 change?

ikreymer replied:
yes, definitely, perhaps should merge that first then..

ikreymer replied:
Refactored to get_host_endpoint_url() to compute this

tw4l reviewed (commented):
Overall looks good, a few questions

ikreymer committed:
refactor from sig v4:
  - add get_host_endpoint_url() to get specific host endpoint
  - only create config for presigning

ikreymer committed:
don't assume all files in an org have same storage: group by stroageref for bulk presigning

ikreymer committed:
add CollectionAllResponse

tw4l reviewed (approved):
Should test a bit more on dev after the latest changes but looks good!

ikreymer committed:
fix missing key in host_endpoint_url

ikreymer commented:
> Should test a bit more on dev after the latest changes but looks good!

Good suggestions! Found and fixed a typo, now should be good!
